File: archive/pulumi-ts/Pr5516/lib/security-stack.ts
import * as pulumi from '@pulumi/pulumi';
import * as aws from '@pulumi/aws';

export interface SecurityStackArgs {
  environmentSuffix: string;
  vpcId: pulumi.Input<string>;
  tags: pulumi.Input<{ [key: string]: string }>;
}

export class SecurityStack extends pulumi.ComponentResource {
  public readonly albSecurityGroup: aws.ec2.SecurityGroup;
  public readonly instanceSecurityGroup: aws.ec2.SecurityGroup;

  constructor(
    name: string,
    args: SecurityStackArgs,
    opts?: pulumi.ComponentResourceOptions
  ) {
    super('tap:security:SecurityStack', name, args, opts);

    // ALB Security Group
    this.albSecurityGroup = new aws.ec2.SecurityGroup(
      `alb-sg-${args.environmentSuffix}`,
      {
        vpcId: args.vpcId,
        description: 'Security group for Application Load Balancer',
        ingress: [
          {
            protocol: 'tcp',
            fromPort: 443,
            toPort: 443,
            cidrBlocks: ['0.0.0.0/0'],
            description: 'Allow HTTPS traffic from internet',
          },
          {
            protocol: 'tcp',
            fromPort: 80,
            toPort: 80,
            cidrBlocks: ['0.0.0.0/0'],
            description: 'Allow HTTP traffic from internet',
          },
        ],
        egress: [
          {
            protocol: '-1',
            fromPort: 0,
            toPort: 0,
            cidrBlocks: ['0.0.0.0/0'],
            description: 'Allow all outbound traffic',
          },
        ],
        tags: {
          ...pulumi.output(args.tags).apply(t => t),
          Name: `alb-sg-${args.environmentSuffix}`,
        },
      },
      { parent: this }
    );

    // Instance Security Group
    this.instanceSecurityGroup = new aws.ec2.SecurityGroup(
      `instance-sg-${args.environmentSuffix}`,
      {
        vpcId: args.vpcId,
        description: 'Security group for EC2 instances',
        ingress: [
          {
            protocol: 'tcp',
            fromPort: 443,
            toPort: 443,
            securityGroups: [this.albSecurityGroup.id],
            description: 'Allow HTTPS traffic from ALB',
          },
          {
            protocol: 'tcp',
            fromPort: 80,
            toPort: 80,
            securityGroups: [this.albSecurityGroup.id],
            description: 'Allow HTTP traffic from ALB for health checks',
          },
        ],
        egress: [
          {
            protocol: '-1',
            fromPort: 0,
            toPort: 0,
            cidrBlocks: ['0.0.0.0/0'],
            description: 'Allow all outbound traffic',
          },
        ],
        tags: {
          ...pulumi.output(args.tags).apply(t => t),
          Name: `instance-sg-${args.environmentSuffix}`,
        },
      },
      { parent: this }
    );

    this.registerOutputs({
      albSecurityGroupId: this.albSecurityGroup.id,
      instanceSecurityGroupId: this.instanceSecurityGroup.id,
    });
  }
}

File: archive/pulumi-java/Pr2276/lib/src/main/java/app/components/SecurityStack.java
package app.components;

import app.config.AppConfig;
import app.utils.TagUtils;
import com.pulumi.aws.ec2.SecurityGroup;
import com.pulumi.aws.ec2.SecurityGroupArgs;
import com.pulumi.aws.ec2.inputs.SecurityGroupEgressArgs;
import com.pulumi.aws.ec2.inputs.SecurityGroupIngressArgs;
import com.pulumi.core.Output;
import com.pulumi.resources.ComponentResource;
import com.pulumi.resources.ComponentResourceOptions;
import com.pulumi.resources.CustomResourceOptions;

import java.util.List;

public class SecurityStack extends ComponentResource {
    private final Output<String> webSecurityGroupId;

    public SecurityStack(final String name, final Output<String> vpcId, final AppConfig config, final ComponentResourceOptions options) {
        super("custom:infrastructure:SecurityStack", name, options);

        // Create Security Group for Web Servers
        var webSecurityGroup = new SecurityGroup(name + "web-security-group",
                SecurityGroupArgs.builder()
                        .name("web-security-group")
                        .description("Security group for web servers")
                        .vpcId(vpcId)
                        .ingress(List.of(
                                SecurityGroupIngressArgs.builder()
                                        .description("HTTP")
                                        .fromPort(80)
                                        .toPort(80)
                                        .protocol("tcp")
                                        .cidrBlocks("0.0.0.0/0")
                                        .build(),
                                SecurityGroupIngressArgs.builder()
                                        .description("HTTPS")
                                        .fromPort(443)
                                        .toPort(443)
                                        .protocol("tcp")
                                        .cidrBlocks("0.0.0.0/0")
                                        .build(),
                                SecurityGroupIngressArgs.builder()
                                        .description("SSH")
                                        .fromPort(22)
                                        .toPort(22)
                                        .protocol("tcp")
                                        .cidrBlocks("0.0.0.0/0")
                                        .build()
                        ))
                        .egress(SecurityGroupEgressArgs.builder()
                                .fromPort(0)
                                .toPort(0)
                                .protocol("-1")
                                .cidrBlocks("0.0.0.0/0")
                                .build())
                        .tags(TagUtils.getTagsWithName("Web-Security-Group", config))
                        .build(), CustomResourceOptions.builder()
                .parent(this)
                .build());

        this.webSecurityGroupId = webSecurityGroup.id();
    }

    public Output<String> getWebSecurityGroupId() {
        return webSecurityGroupId;
    }
}
